Bearded.UI/Events/EventRouter.cs

```diff
@@ -42,12 +36,60 @@ public static EventPropagationPath FindPropagationPath(IControlParent root, Cont
                     return EventPropagationPath.Empty;
                 }
                 path.Add(parentAsControl);
-                
+
                 parent = parentAsControl.Parent;
             }
 
             path.Reverse();
             return new EventPropagationPath(path.AsReadOnly());
         }
+
+        private static IEnumerable<Control> findPropagationPathUsingTest(
+            IControlParent root, PropagationTest propagationTest)
+        {
+            var path = new List<Control>();
+            findPropagationPathUsingTest(root, propagationTest, path);
+            return path;
+        }
+
+        private static bool findPropagationPathUsingTest(
+            IControlParent root, PropagationTest propagationTest, List<Control> path)
+        {
+            foreach (var child in root.Children.Reverse())
+            {
+                var outcome = propagationTest(child);
+                switch (outcome)
+                {
+                    case PropagationTestOutcome.Miss:
+                        break;
+                    case PropagationTestOutcome.PassThrough:
+                        if (child is IControlParent childAsPassThroughParent
+                            && childAsPassThroughParent.Children.Count > 0)
+                        {
+                            path.Add(child);
+                            var hasPathWithHit =
+                                findPropagationPathUsingTest(childAsPassThroughParent, propagationTest, path);
+                            if (hasPathWithHit)
+                            {
+                                return true;
+                            }
+                            // backtrack
+                            path.RemoveAt(path.Count - 1);
```

Comment:
It bothers me a little that it's doing this piece of code for every step if it has to backtrack. But I'm not sure there's an easy and clear way to prevent that.

Reply:
What I mean is, in principle, you could remove a whole range in one call at the top of the pass-through path. But that would require passing extra information around and would probably not be worth the trouble. Not to mention that we wouldn't expect there to be large trees with complex hierarchies of lots of pass-through controls, would we? 🤔

Reply:
I would hope not. I also think in this case the readability argument would win. We can revisit if performance ever becomes an issue, but I would expect this algorithm to be run once per frame, so it's probably fine.

Reply:
Yes, this will never be the bottleneck, unless you start using this system to do actual game logic raytracing or something.